### PR TITLE
tpnote: init at 1.23.9

### PR DESCRIPTION
**Pull request by getreu**
## Add new package: Tp-Note

**Tp-Note: Markup enhanced granular note-taking**

*Save and edit your clipboard content as a note file*

[Cargo](
https://crates.io/crates/tp-note)
[Documentation](
https://docs.rs/tpnote-lib)
[License](
https://gitlab.com/getreu/tp-note)

_Tp-Note_ is a note-taking-tool and a template system - freely available for
Windows, MacOS and Linux - that consistently synchronizes the note’s meta-data
with its filename. _Tp-Note_'s main design goal is to convert some input text -
usually provided by the system's clipboard - into a Markdown note file with
a descriptive YAML header and a meaningful filename.
_Tp-Note_ collects various information about its environment
and the clipboard and stores them in variables. New notes are created by
filling these variables in predefined and customizable _Tera_-templates.
_TP-Note's_ default templates are written in Markdown and can be easily adapted
to any other markup language if needed. After creating a new note, _TP-Note_
launches the system's text editor and connects the default web browser to _Tp-
Note_'s internal Markdown/RestructuredText renderer and web server. The viewer
detects note file changes and updates the rendition accordingly.

_On Tue, 2023-12-19 at 12:58 +1100, Dev Rain wrote:_

> _Found Tp-Note awhile back and it has become part of my daily workflow,
> and indeed part of my daily note-taking life. I wanted to extend my
> thanks; so thank you. 
> dev.rain_

Read more in [Tp-Note’s user manual], [Download Tp-Note] or visit the project 
page: [Tp-Note - Minimalistic note-taking].

[Tp-Note’s user manual]: https://blog.getreu.net/projects/tp-note/tpnote--manual.html
[Download Tp-Note]: https://blog.getreu.net/projects/tp-note/index.html#distribution
[Tp-Note - Minimalistic note-taking]: https://blog.getreu.net/projects/tp-note/

## Things done

- Builts on platform(s) (tested with `cargo test` and `cargo build`) 
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin

- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [-] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage) [does not apply]
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by patka-123 — CHANGES_REQUESTED** [review]

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

Please make sure to read the contributing guide to know how to format your commit message and the PR title. The current state is incorrect.
Once it's done, I'll review it once again, in the meantime, I'll set the PR to draft, remove it from draft when it's done.

**Comment by getreu** [comment]
> Would you be able to add yourself to the maintainer list? You can read more about how to do that [here](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md#how-to-become-a-maintainer). You added yourself as a maintainer for the package so it also needs an addition to the mainainer list.

Done: https://github.com/NixOS/nixpkgs/pull/294942

**Review by drupol — CHANGES_REQUESTED** [review]
Thanks for making the changes.

This PR should only contains 2 commits only.

1. One commit to add you as maintainer: `maintainer: add getreu`
2. One commit to add `tpnote`.

File: pkgs/by-name/tp/tpnote/package.nix
Comment (drupol) [inline pkgs/by-name/tp/tpnote/package.nix:35]:
```suggestion
  ];
```

File: pkgs/by-name/tp/tpnote/package.nix
Comment (drupol) [inline pkgs/by-name/tp/tpnote/package.nix:46]: Consider removing `with lib;`, sorting meta attributes alphabetically and add `meta.mainProgram`.

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/by-name/tp/tpnote/package.nix
Comment (drupol) [inline pkgs/by-name/tp/tpnote/package.nix:53]:
```suggestion
  meta = {
    changelog = "https://github.com/getreu/tp-note/releases/tag/v${version}";
    description = "Markup enhanced granular note-taking";
    homepage = "https://blog.getreu.net/projects/tp-note/";
    license = lib.licenses.mit;
    mainProgram = "tpnote";
    maintainers = with lib.maintainers; [ getreu ];
  };
```

**Review by drupol — CHANGES_REQUESTED** [review]
Here's the solution. This really highlight how `with lib;` is wrongly understood and why it is a good idea to remove it!

**Comment by getreu** [comment]
@drupol Thank you for your help.

**Review by drupol — APPROVED** [review]

**Review by patka-123 — APPROVED** [review]

**Comment by drupol** [comment]
I'm just waiting for at least one build on Darwin before merging.

**Review by drupol — CHANGES_REQUESTED** [review]
There's a building issue on Darwin:

```
       >   = note: ld: framework not found CoreServices
```

Do you think you could fix it?

Hint: Search for `CoreServices` in `nixpkgs` to find some code inspiration.

**Comment by getreu** [comment]
I found this:

```nix
 ++ lib.optionals stdenv.isDarwin (with darwin.apple_sdk.frameworks; [
    CoreServices SystemConfiguration
  ]);
```

Does this help? How can I test?

**Comment by drupol** [comment]
The issue is different now! That's a good sign.

```
       >   = note: ld: framework not found AppKit
```

**Comment by drupol** [comment]
Moving the PR to draft until the builds on Darwin are fixed.

**Comment by getreu** [comment]
>   = note: ld: framework not found AppKit

I found these links, but today I have not time to dig further into it. 
* [NixOS Search - Packages - AppKit](https://search.nixos.org/packages?channel=unstable&show=gnustep.base&from=0&size=50&sort=relevance&type=packages&query=AppKit)
* [How to link AppKit framework in rustflags? - help - The Rust Programming Language Forum](https://users.rust-lang.org/t/how-to-link-appkit-framework-in-rustflags/102877)

I have experience with Darwin. Where can I ask? Help is welcome.

**Comment by drupol** [comment]
> The issue is different now! That's a good sign.
> 
> ```
>        >   = note: ld: framework not found AppKit
> ```

The solution is here, just add `AppKit` in the dependencies, just like you did for `CoreServices`.

Hint: Look for the keyword `AppKit` in `nixpkgs`.

File: pkgs/by-name/tp/tpnote/package.nix
Comment (getreu) [reply pkgs/by-name/tp/tpnote/package.nix:40]: Ok, should it be simply: 
```nix
++ lib.optionals stdenv.isDarwin (with darwin.apple_sdk.frameworks; [
    CoreServices SystemConfiguration AppKit
  ]);
```
?

Comment (drupol) [reply pkgs/by-name/tp/tpnote/package.nix:40]: Let's follow the trend in `nixpkgs`:

```nix
++ lib.optionals stdenv.isDarwin (with darwin.apple_sdk.frameworks; [
     AppKit
     CoreServices
     SystemConfiguration
  ]);
```

Comment (getreu) [reply pkgs/by-name/tp/tpnote/package.nix:40]: I rollback my last try and give a new chance tomorrow. Thanks for your feedback so far.

Comment (getreu) [reply pkgs/by-name/tp/tpnote/package.nix]: I did not see your answer. I'll try that right now.

Comment (getreu) [reply pkgs/by-name/tp/tpnote/package.nix]: All checks have finally passed. Thank you for your help.

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
Thanks for your first contribution, welcome to the nix family!

Keep PRs coming!

**Pull request merged**
